src/FSharp.Core/array.fs

```diff
@@ -1936,6 +1936,214 @@ module Array =
+    let randomChoicesWith (random: Random) (count: int) (source: 'T array) : 'T array =
+        checkNonNull "source" source
+        checkNonNull "random" random
+
+        if count < 0 then
+            invalidArgInputMustBeNonNegative "count" count
+
+        let inputLength = source.Length
+
+        if inputLength = 0 then
+            invalidArg "source" LanguagePrimitives.ErrorStrings.InputArrayEmptyString
+
+        let result = Microsoft.FSharp.Primitives.Basics.Array.zeroCreateUnchecked count
+
+        for i = 0 to count - 1 do
+            let j = random.Next(0, inputLength)
+            result[i] <- source[j]
+
+        result
+
+    [<CompiledName("RandomChoicesBy")>]
+    let randomChoicesBy (randomizer: unit -> float) (count: int) (source: 'T array) : 'T array =
+        checkNonNull "source" source
+
+        if count < 0 then
+            invalidArgInputMustBeNonNegative "count" count
+
+        let inputLength = source.Length
+
+        if inputLength = 0 then
+            invalidArg "source" LanguagePrimitives.ErrorStrings.InputArrayEmptyString
+
+        let result = Microsoft.FSharp.Primitives.Basics.Array.zeroCreateUnchecked count
+
+        for i = 0 to count - 1 do
+            let j = Microsoft.FSharp.Primitives.Basics.Random.next randomizer 0 inputLength
+            result[i] <- source[j]
+
+        result
+
+    [<CompiledName("RandomChoices")>]
+    let randomChoices (count: int) (source: 'T array) : 'T array =
+        randomChoicesWith ThreadSafeRandom.Shared count source
+
+    [<CompiledName("RandomSampleWith")>]
+    let randomSampleWith (random: Random) (count: int) (source: 'T array) : 'T array =
+        checkNonNull "source" source
+        checkNonNull "random" random
+
+        if count < 0 then
+            invalidArgInputMustBeNonNegative "count" count
+
+        let inputLength = source.Length
+
+        if inputLength = 0 then
+            invalidArg "source" LanguagePrimitives.ErrorStrings.InputArrayEmptyString
+
+        if count > inputLength then
+            invalidArgOutOfRange "count" count "source.Length" inputLength
+
+        let result = Microsoft.FSharp.Primitives.Basics.Array.zeroCreateUnchecked count
+
+        // algorithm taken from https://github.com/python/cpython/blob/main/Lib/random.py
+        let setSize =
+            Microsoft.FSharp.Primitives.Basics.Random.getMaxSetSizeForSampling count
+
+        if inputLength <= setSize then
+            let pool = copy source
+
+            for i = 0 to count - 1 do
+                let j = random.Next(0, inputLength - i)
+                result[i] <- pool[j]
+                pool[j] <- pool[inputLength - i - 1]
+        else
+            let selected = HashSet()
+
+            for i = 0 to count - 1 do
+                let mutable j = random.Next(0, inputLength)
+
+                while selected.Contains(j) do
+                    j <- random.Next(0, inputLength)
+
+                selected.Add(j) |> ignore
```

Comment:
Here and elsewhere (it should halve the number of lookup operations):

```suggestion
                while not (selected.Add j) do
                    j <- random.Next(0, inputLength)
```

Reply:
Nice observation! Fixed